lib/flipper/adapters/active_record.rb

```diff
@@ -244,17 +245,26 @@ def delete(feature, gate)
       end
 
       def enable_multi(feature, gate, thing)
-        with_connection(@gate_class) do
-          @gate_class.create! do |g|
-            g.feature_key = feature.key
-            g.key = gate.key
-            g.value = thing.value.to_s
+        with_connection(@gate_class) do |connection|
+          begin
+            connection.transaction do
+              savepoint_name = "flipper_savepoint_#{SecureRandom.hex(8)}"
+              connection.execute("SAVEPOINT #{savepoint_name}")
```

Comment:
Noting that: pg transaction `savepoints` can cause performance degradation on postgres. 

Gitlab have blogged about this and how they remove them from their app https://about.gitlab.com/blog/2021/09/29/why-we-spent-the-last-month-eliminating-postgresql-subtransactions/

Reply:
@camallen yep, I'm aware. This is a very low throughput method. Even at a large company/app I'd be surprised if it was executed more than 100 times in a day. I think that level of throughput this should be fine. Knowing that, are you still worried? 

If anyone knows another way to fix it, I'm all ears but after hacking around on it for 30 minutes this was the first thing that worked.

Reply:
> I think that level of throughput this should be fine. Knowing that, are you still worried?

I am still a little worried about this change, perhaps unduly.

From my reading the gitlab investigation showed that any savepoint use on the primary can impact all replicas as the replica has to record all active transactions after the use of a savepoint and this tracking is a problem during any 'long' running queries. On a very active system the duration of a 'long' query will vary as it depends on the number of active transactons. If there are enough they will saturate the pg server cache and spill from memory to disk and then impact query perf due to slower IO on hot code paths. 

My concern is around the the unpredictable nature of this issue due to the interplay of the active transaction count and the duration of any queries that impact the replicas. It'll appear like random performance degradation that (hopefully) eventually self heals after the queries complete. 

Teams that use replicas most likely are already experiencing a high transaction load on the primary and this change then might open them up to this situation.

One of the [proposed solutions](https://postgres.ai/blog/20210831-postgresql-subtransactions-considered-harmful#ideas-for-postgresql-development) (that may have landed in new pg version) was to allow a configurable PG server cache size to avoid spilling over to disk and slow IO on a hot code path. However by default that isn't available on older pg versions and may not be configurable on pg cloud variants. 

Again i may be reading it wrong, I'm wary as I've been bitten in the past from long running queries impacting replica query performance and the random nature of these issues.

> If anyone knows another way to fix it, I'm all ears but after hacking around on it for 30 minutes this was the first thing that worked.

ARs postgres adapter has a private [in_transaction?](https://github.com/rails/rails/blob/93e23dee7a2226e69bf44a5ba18bae445bd36399/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L899-L901) method to detect if we're in a transaction that uses `open_transactions`. We might be able to introspect on the connection adapter's [open_transactions](https://github.com/rails/rails/blob/93e23dee7a2226e69bf44a5ba18bae445bd36399/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L367) size to determine if we're in a transaction and if we need to use a savepoint sql directive instead of opting everyone into the use of savepoints by default. 

I'm not privy if there is a better way to introspect the AR connection transaction stack state to determine if we can 'opt' in to the use nested transactions. Maybe @byroot has thoughts.

Reply:
> Maybe @byroot has thoughts.

I don't know the code, so hard. But it looks like it's trying to insert a record and ignore if the record already exist? If so then, `ON DUPLICATE KEY IGNORE` wouldn't require a sub transaction.

Reply:
Ref to similar discussion when we used a savepoint in find_or_create https://github.com/rails/rails/issues/51052#issuecomment-1965659452.

Reply:
@camallen this isn't a hot code path and shouldn't ever be. 

If it is, then you aren't using flipper correctly. Individual actors is for enabling a feature for a few people here and there per feature. Anything that requires 100+ should be in a group or % rollout as detailed in the docs. 

Any write in flipper is a rare thing compared to reads and that is what flipper is optimized for.

Reply:
> this isn't a hot code path and shouldn't ever be.

Apologies for the confusion, when I used the 'hot code path' term, I was referring to the PostgreSQL server when a `savepoint` is used on a primary and the replica then has to start tracking all know transactions. I wasn't referring to flipper configuring a feature flag, though flipper could be the trigger when it issues a `savepoint` transaction. 

I still think there is a risk with this change. If the GitLab investigation is correct then this change introduces a ticking bomb that may at some point explode for busy PostgreSQL systems that have replicas enabled. Similar to the concerns raised in https://github.com/rails/rails/issues/51052

Would you be open to a PR that attempts to use the upsert on conflict do nothing approach that byroot mentioned? I'd be up for working on this if it would be considered to land.

A naive implementation would run the model validations and if the model is valid it would then use ARs `insert` / `insert_all` [methods](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-insert_all) combined with :unique_by to skip the duplicate records and thus not poison the outer transaction. 

A caveat being that the `insert on conflict` feature is only available on [PG versions 9.5+](https://www.postgresql.org/docs/9.5/sql-insert.html) and active record [supports PG 9.3+ versions ](https://github.com/rails/rails/blob/0952b2a0a2a741dc6d1b873890d1b8f1210ee9cd/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L661-L662). It appears that flipper supports [active record 4.2+ ](https://github.com/flippercloud/flipper/blob/4b911dd3d185b5ba8d7f0be1323a5ba633565af1/flipper-active_record.gemspec#L35) so any implementation would require branching code based on the servers feature availability [detects server feature availability](https://github.com/rails/rails/blob/0952b2a0a2a741dc6d1b873890d1b8f1210ee9cd/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L268-L273). 

If the server doesn't support `ON CONFLICT DO NOTHING` SQL then the current `requires_new` savepoint would be used.

Reply:
Looks like the configurable cache sizes have landed in PG 17 https://pganalyze.com/blog/5mins-postgres-17-configurable-slru-cache

Reply:
@camallen happy to take a look at a PR. Again, this is super low throughput so I don't believe it'll be an issue.